File: CadenaCaracteres.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int afd(char *str) {
    int estado = 0;
    int i = 0;

    while (str[i] != '\0' && str[i] != '\n') {
        char c = str[i];
        switch (estado) {

            case 0:
                if (isalpha((unsigned char)c))       estado = 1;
                else if (isdigit((unsigned char)c))  estado = 2;
                else if (c == '+')                   estado = 3;
                else return 0;
                break;
            case 1:

                if (islower((unsigned char)c))       estado = 5;
                else return 0;
                break;
            case 2:

                if (!isdigit((unsigned char)c)) return 0;
                break;
            case 3:

                if (c == '+') estado = 4;
                else return 0;
                break;
            case 4:

                return 0;
            case 5:

                if (isdigit((unsigned char)c)) estado = 6;
                else return 0;
                break;
            case 6:

                if (islower((unsigned char)c)) estado = 5;
                else return 0;
                break;
        }
        
        i++;
    }
    return (estado == 1 || estado == 6 || estado == 2 || estado == 3 || estado == 4);
}

int main() {
    FILE *archivo = fopen("archivo.txt", "r");
    if (!archivo) {
        printf("Error al abrir el archivo.\n");
        return 1;
    }

    char palabra[512];
    while (fscanf(archivo, "%511s", palabra) != EOF) {
        if (afd(palabra)) printf("%s: ACEPTA\n", palabra);
        else printf("%s: NO ACEPTA\n", palabra);
    }

    fclose(archivo);
    return 0;
}

File: CadenaCaracteres.py
import sys

def afd_scanner(lexema):
  
    estado = 0
    
    for char in lexema:
        if estado == 0:
            if char.isalpha(): estado = 1
            elif char.isdigit(): estado = 2
            elif char == '+': estado = 3
            else: return False
        
        elif estado == 1: 
            if char.isalnum(): estado = 1
            else: return False
            
        elif estado == 2:
            if char.isdigit(): estado = 2
            else: return False
            
        elif estado == 3: 
            if char == '+': estado = 4
            else: return False 
            
        elif estado == 4: 
            return False
            
    return estado in [1, 2, 3, 4]

def procesar_archivo(nombre_archivo):
    try:
        with open(nombre_archivo, 'r') as f:
            for linea in f:
                palabras = linea.split()
                for p in palabras:
                    resultado = "ACEPTA" if afd_scanner(p) else "NO ACEPTA"
                    print(f"{p}: {resultado}")
    except FileNotFoundError:
        print("Archivo no encontrado.")

if __name__ == "__main__":
    procesar_archivo("archivo.txt")
